ADR/0039-workspace-deprecation.md

```diff
@@ -0,0 +1,98 @@
+
+In order to support multi-cluster deployment, [kubesaw] was used instead of KCP.
+
+Today, Workspaces are an important concept in the Konflux user experience. The user onboards into a workspace of their own. A team member of theirs may add them to a team workspace. As a platform administrator, if you hear about a user having trouble, you'll ask them what workspace they're working in.
+
+A Workspace is an abstraction of a namespace. Each workspace has a single namespace
+which backs it.
+
+Workspaces are required by [kubesaw] for providing a transparent multi-cluster communication for us. A user's workspace may be allocated to one cluster or another - but our proxy layer keeps the user blissfully unaware of this. Workspaces can be placed on clusters either automatically (using [kubesaw] mechanisms) or manually by a platform administrator.
+
+Workspace roles give a way to describe who has what rights in the workspace.
+
+Workspaces happen to be our mechanism for managing quota - a fact inherited from the [kubesaw] stack.
+
+### Some changes
+
+Some relevant changes that have occurred or are ongoing.
+
+* We want to make Konflux a successful open source projects that fits into the
+cloud native apps ecosystem. We want everything to work on vanilla Kubernetes - no OpenShift required.
+
+* We had our [first upstream Konflux release](https://github.com/konflux-ci/konflux-ci) which doesn't use [kubesaw] or relying
+on the Workspace abstraction, and is deployable on vanilla Kubernetes.
+
+* We made a decision (now long ago) to drop [KCP] from our architecture, which initially introduced the workspace concept. Both transparent multi-cluster as a feature and workspaces as a concept came from KCP.
+
+* We made the decision (at Red Hat) to stop trying to offer a managed service interface for Konflux. The user signup, waitlist, and workspace provisioning interfaces made a lot of sense there.
+
+## Decision
+
+1. We will remove the Workspace abstraction from Konflux and use standard Kubernetes namespaces
+directly for providing a place for users to run their pipelines and store their configurations
+and secrets.
+
+2. We will stop using the term Workspace, and start to use the term Namespace.
+
+3. The Konflux UI will expose a wizard for creating a new namespace. This
+wizard will be available to any user that has permissions to create namespaces.
+
+4. Konflux will provide an optional automation for creating a namespace for a user
+when he/she logins for the first time.
+
+5. Konflux won't be opinionated about the mechanism for initializing/maintaining namespaces with
+supporting resources such as ResourceQuota and LimitRange. Konflux will differ
+this responsibility to other tools which are specialized in this task such as 
+(but not limited to) [Kyverno](https://kyverno.io/policies/best-practices/add-ns-quota/add-ns-quota/) and [ArgoCD](https://github.com/konflux-ci/namespace-generator).
```

Comment:
How many other tools do we perceive would need to be added if we remove the Workspace abstraction?

Do we know that users/platform engineers of Konflux would _want_ to use these various tools instead? Do we know that the platform engineers would have sufficient subject matter expertise on these tools so that there would be no additional ramp-up on them (i.e. so that they wouldn't have to learn about multiple tools instead of just the one controller)?

Would it be possible to create a workspace implementation which instead leveraged these same tools under the cover to allow platform engineers to make this decision if they would prefer?

Reply:
> How many other tools do we perceive would need to be added if we remove the Workspace abstraction?
depends on the features the user want. 

> Do we know that users/platform engineers of Konflux would want to use these various tools instead? Do we know that the platform engineers would have sufficient subject matter expertise on these tools so that there would be no additional ramp-up on them (i.e. so that they wouldn't have to learn about multiple tools instead of just the one controller)?

1. The tools that I mentioned are just an example, maybe the user will have other ways. It's not a requirement to use none of them, Konflux can work without ResourceQuotas, LimitRanges, policy for managing namespaces etc.

2. Even if we will duplicate the functionality of those tools into Konflux, the user will need to learn how to use our own implementation (or abstraction).

3. The tools I mentioned are widely used in the k8s ecosystem, two of them are graduated CNCF projects. I don't think it would be an issue for a Konflux operator to understand how to use them. I also think that showing that Konflux can be used with those tools will help Konflux find its place in the k8s ecosystem.

> Would it be possible to create a workspace implementation which instead leveraged these same tools under the cover to allow platform engineers to make this decision if they would prefer?

To me it feels like a grantee for having a leaky abstraction and duplication of work.

Reply:
If the are no objections I'll close this thread.